hibernate-envers/src/main/java/org/hibernate/envers/query/criteria/internal/EntityTypeAuditExpression.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later.
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.envers.query.criteria.internal;
+
+import java.util.Map;
+
+import org.hibernate.envers.boot.internal.EnversService;
+import org.hibernate.envers.internal.reader.AuditReaderImplementor;
+import org.hibernate.envers.internal.tools.query.Parameters;
+import org.hibernate.envers.internal.tools.query.QueryBuilder;
+import org.hibernate.envers.query.criteria.AuditCriterion;
+
+/**
+ * @author Felix Feisst (feisst dot felix at gmail dot com)
+ */
+public class EntityTypeAuditExpression implements AuditCriterion {
+
+	private String alias;
+	private String entityName;
+
+	public EntityTypeAuditExpression(
+			String alias,
+			String entityName) {
+		this.alias = alias;
+		this.entityName = entityName;
+	}
+
+	@Override
+	public void addToQuery(
+			EnversService enversService,
+			AuditReaderImplementor auditReader,
+			Map<String, String> aliasToEntityNameMap,
+			String baseAlias,
+			QueryBuilder qb,
+			Parameters parameters) {
+		String effectiveAlias = alias == null ? baseAlias : alias;
+		String effectiveEntityName = entityName;
+		if ( enversService.getEntitiesConfigurations().isVersioned( effectiveEntityName ) ) {
```

Comment:
Would you not agree that it makes sense to throw a `NotAuditedException` if determined that the `effectiveEntityName` is not a versioned entity?

Reply:
I was looking back at this a bit today and I believe we can add this check without any API change.  I believe what should work here would be to modify `AbstractAuditQuery` as follows:
```
public AuditQuery add(AuditCriterion criterion) {
	validateSubTypeCriterionOnQueryRoot( criterion );
	criterions.add( criterion );
	return this;
}
```
Then the validate check is as follows:
```
private void validateSubTypeCriterionOnQueryRoot(AuditCriterion criterion) {
	if ( criterion instanceof EntityTypeAuditExpression ) {
		EntityTypeAuditExpression expression = (EntityTypeAuditExpression) criterion;
		final String entityTypeName = expression.getEntityName();
		if ( !enversService.getEntitiesConfigurations().isVersioned( entityTypeName ) ) {
			throw new AuditException( "Cannot apply sub-type criterion for non-audited entity " + entityTypeName );
		}
	}
}
```
The test case is where `EntityA` is audited and `EntityB` extends `EntityA`; however `EntityB` is not audited.  So the following query fails:
```
@Test
public void testQuerySubTypeNonAuditedType() {
	// This performs a sub-type predicate using a non-audited type on the query root.
	// In this use case, the query should not execute and should throw an exception.
	try {
		List<?> list = getAuditReader().createQuery().forEntitiesAtRevision( EntityA.class, 1 )
				.add( AuditEntity.entityType( EntityB.class ) )
				.addProjection( AuditEntity.property( "name" ) )
				.addOrder( AuditEntity.property( "name" ).asc() )
				.getResultList();
	}
	catch ( Exception e ) {
		assertTyping( AuditException.class, e );
	}
}
```
However I also have a `NOT_AUDITED` association from `EntityA` to `EntityX` where `EntityX` also has a sub-type of `EntityY`.  In this use case, I can apply the sub-type criterion just fine without the exception being triggered:
```
@Test
public void testRestrictToSubTypeOnTraversal1() {
	List<?> list = getAuditReader().createQuery().forEntitiesAtRevision( EntityA.class, 1 )
			.traverseRelation( "x", JoinType.LEFT )
			.add( AuditEntity.entityType( EntityX.class ) )
			.up()
			.addProjection( AuditEntity.property( "name" ) )
			.addOrder( AuditEntity.property( "name" ).asc() )
			.getResultList();
	assertEquals( "Expect all EntityA types that are associated with EntityX types.", list( "a1" ), list );
}

@Test
public void testRestrictToSubTypeOnTraversal2() {
	List<?> list = getAuditReader().createQuery().forEntitiesAtRevision( EntityA.class, 1 )
			.traverseRelation( "x", JoinType.LEFT )
			.add( AuditEntity.entityType( EntityY.class ) )
			.up()
			.addProjection( AuditEntity.property( "name" ) )
			.addOrder( AuditEntity.property( "name" ).asc() )
			.getResultList();
	assertEquals( "Expect all EntityA types that are associated with EntityY types.", list( "a2" ), list );
}
```
I believe this satisfies all situations here, what do you think @ffeisst ?

Reply:
Nice work! I also believe that this satisfies all situations. Fail-early is definitely a good thing, especially for the Envers Query API which could easily be used wrong leading to non-meaningful error messages. These assertions are required at a hundred other places, too. Step by step.